File: src/main/scala/org/tindalos/principle/domain/analyzers/structure/Graph.scala
package org.tindalos.principle.domain.analyzers.structure

import scala.collection.JavaConverters._

object Graph {

  //def nodeById(id: String) = graph.find(_.id == id).get
  //def findDirectDownstreamNodes(n: Node, graph: Set[Node]) = n.dependencies.map(nodeById)
  //def findDirectUpstreamNodes(n: Node, graph: Set[Node]) = n.dependants.map(nodeById)

  def isValid(graph: Set[Node]): Boolean = {

    def symmetryHold(pair: (Node, Node)) =
      if (pair._1.dependencies.contains(pair._2.id)) //if A->B
        (pair._2.dependants.contains(pair._1.id)) // then B has A as dependant
      else if (pair._1.dependants.contains(pair._2.id)) //if A<-B
        (pair._2.dependencies.contains(pair._1.id)) // then
      else true // no relationship

    val pairs = for {
      n1 <- graph
      n2 <- graph if n1 != n2
    } yield (n1, n2)

    pairs.forall(symmetryHold)
  }

  def isIsland(subgraph: Set[Node]) = {
    val subgraphDependencies = subgraph.flatMap(n => n.dependants.asScala ++ n.dependants.asScala)
    val externalDependencies = subgraphDependencies &~ subgraph.map(_.id)
    externalDependencies.isEmpty
  }

  def findDownstreamNodes(n: Node, graph: Set[Node]) = {

    val nodeMap: Map[String, Node] = graph.groupBy(_.id).map(kv => (kv._1, kv._2.toList.head))

    def helper(node: Node, acc: Set[Node]): Set[Node] = {

      val newAcc = acc + node
      val nextNodes = node.dependencies.asScala.flatMap(id => nodeMap.get(id)).toSet -- newAcc

      // recursively call on dependencies
      newAcc ++ nextNodes.flatMap(x => helper(x, newAcc ++ nextNodes))
    }
    helper(n, Set())
  }

  def findSources(graph: Set[Node]): Set[Node] = graph.filter(n => n.dependants.isEmpty)

  private def findDetachableSubgraph(n: Node, graph: Set[Node]) = {
    def helper(startNode: Node, upstreamNodes: Set[Node]): Set[Node] = {
      val result = for {
        node <- graph
        if (startNode.dependencies.contains(node.id) //node is a direct dependency of startNode
          && ((node.dependants.asScala - startNode.id) &~ upstreamNodes.map(_.id)).isEmpty //node has no upstream deps outside upstreamNodes
          && !upstreamNodes.contains(node)) //and node is not among upstreamNodes
      } yield helper(node, upstreamNodes + startNode)
      (upstreamNodes + startNode) ++ result.flatten
    }
    helper(n, Set())
  }

  def findDetachableSubgraphs(graph: Set[Node]): SubgraphDecomposition = {
    val result = for {
      node <- graph
      sg = Graph.findDetachableSubgraph(node, graph)
      if (sg.size > 1)
    } yield (node, sg)

    val peninsulas = result
      .groupBy(_._2) //Map[Set[Node],Set[(Node,Set[Node])]]
      .toList
      .map { kv =>
        val subgraph: Set[Node] = kv._1
        val frontNodes: Set[Node] = kv._2.map(_._1)
        new Peninsula(frontNodes.asJava, subgraph.asJava)
      }
      .sortBy(_.subgraph.size)
      .reverse

    new SubgraphDecomposition(peninsulas.asJava)
  }

}

File: src/main/java/org/tindalos/principle/domain/analyzers/structure/Graph.java
package org.tindalos.principle.domain.analyzers.structure;

import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Utility operations for navigating and decomposing dependency graphs.
 */
public final class Graph {

    private Graph() {
    }

    public static boolean isValid(Set<Node> graph) {
        for (var first : graph) {
            for (var second : graph) {
                if (!first.equals(second) && !symmetryHolds(first, second)) {
                    return false;
                }
            }
        }
        return true;
    }

    public static boolean isIsland(Set<Node> subgraph) {
        var subgraphIds = subgraph.stream().map(Node::id).collect(Collectors.toSet());
        var externalDependencies = subgraph.stream()
                .flatMap(node -> node.dependants().stream())
                .filter(id -> !subgraphIds.contains(id))
                .collect(Collectors.toSet());
        return externalDependencies.isEmpty();
    }

    public static Set<Node> findDownstreamNodes(Node node, Set<Node> graph) {
        var nodeMap = graph.stream().collect(Collectors.toMap(Node::id, n -> n, (first, second) -> first));
        return findDownstreamNodes(node, Set.of(), nodeMap);
    }

    public static Set<Node> findSources(Set<Node> graph) {
        return graph.stream()
                .filter(node -> node.dependants().isEmpty())
                .collect(Collectors.toUnmodifiableSet());
    }

    public static SubgraphDecomposition findDetachableSubgraphs(Set<Node> graph) {
        var frontNodesBySubgraph = new HashMap<Set<Node>, Set<Node>>();

        for (var node : graph) {
            var subgraph = findDetachableSubgraph(node, graph);
            if (subgraph.size() > 1) {
                frontNodesBySubgraph.computeIfAbsent(subgraph, ignored -> new HashSet<>()).add(node);
            }
        }

        var peninsulas = frontNodesBySubgraph.entrySet().stream()
                .map(entry -> new Peninsula(entry.getValue(), entry.getKey()))
                .sorted(Comparator.comparingInt((Peninsula peninsula) -> peninsula.subgraph().size()).reversed())
                .toList();

        return new SubgraphDecomposition(peninsulas);
    }

    private static boolean symmetryHolds(Node first, Node second) {
        if (first.dependencies().contains(second.id())) {
            return second.dependants().contains(first.id());
        }
        if (first.dependants().contains(second.id())) {
            return second.dependencies().contains(first.id());
        }
        return true;
    }

    private static Set<Node> findDownstreamNodes(Node node, Set<Node> acc, Map<String, Node> nodeMap) {
        var newAcc = new HashSet<>(acc);
        newAcc.add(node);

        var nextNodes = node.dependencies().stream()
                .map(nodeMap::get)
                .filter(Objects::nonNull)
                .collect(Collectors.toCollection(HashSet::new));
        nextNodes.removeAll(newAcc);

        var result = new HashSet<>(newAcc);
        var nextAcc = new HashSet<>(newAcc);
        nextAcc.addAll(nextNodes);
        for (var nextNode : nextNodes) {
            result.addAll(findDownstreamNodes(nextNode, nextAcc, nodeMap));
        }
        return Set.copyOf(result);
    }

    private static Set<Node> findDetachableSubgraph(Node node, Set<Node> graph) {
        return findDetachableSubgraph(node, Set.of(), graph);
    }

    private static Set<Node> findDetachableSubgraph(Node startNode, Set<Node> upstreamNodes, Set<Node> graph) {
        var result = new HashSet<Node>();
        var upstreamIds = upstreamNodes.stream().map(Node::id).collect(Collectors.toSet());

        for (var node : graph) {
            var externalUpstreams = new HashSet<>(node.dependants());
            externalUpstreams.remove(startNode.id());
            externalUpstreams.removeAll(upstreamIds);

            if (startNode.dependencies().contains(node.id())
                    && externalUpstreams.isEmpty()
                    && !upstreamNodes.contains(node)) {
                result.addAll(findDetachableSubgraph(node, append(upstreamNodes, startNode), graph));
            }
        }

        var combined = new HashSet<>(upstreamNodes);
        combined.add(startNode);
        combined.addAll(result);
        return Set.copyOf(combined);
    }

    private static Set<Node> append(Set<Node> nodes, Node node) {
        var result = new HashSet<>(nodes);
        result.add(node);
        return Set.copyOf(result);
    }
}